### PR TITLE
[exporter/elasticsearch] Workaround TSDB array dimension limitation for metrics OTel mode

### LINKED ISSUE #35004: [exporter/elasticsearch] Workaround TSDB limitation on array dimension fields
### Component(s)

exporter/elasticsearch

## Description

Elasticsearch TSDB does not accept dimension fields which are arrays. It will result in ingestion failures. 

We have [a related fix](https://github.com/elastic/elasticsearch/pull/112345) for non-TSDB index mode, but TSDB index mode still has this limitation.

As all attributes are dimensions, and that the exporter currently emits attribute values which are arrays as actual arrays in the document, it will be affected by this TSDB limitation. Elasticsearch exporter should workaround this limitation specifically for metrics.

## Steps to Reproduce

Send an attribute with array value to an ES index that uses TSDB index mode

## Expected Result

No errors

## Actual Result

Fail to index

```
2024-09-04T14:55:22.350+0100    error   elasticsearchexporter@v0.108.0/bulkindexer.go:332       failed to index document        {"kind": "exporter", "data_type": "metrics", "name": "elasticsearch", "index": "metrics-generic.otel-default", "error.type": "illegal_argument_exception", "error.reason": "Error extracting routing: Routing values must be strings but found [START_ARRAY]"}
```

### Collector version

v0.108.0

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by carsonip**
**Description:** <Describe what has changed.>

Elasticsearch TSDB does not support array dimensions. Workaround it by stringifying attribute array values in OTel mapping mode for metrics.

Refactor to improve test code reuse.

**Link to tracking Issue:** <Issue number if applicable>

Fixes #35004

**Testing:** <Describe what testing was performed and which tests were added.>

Added exporter test for otel mode logs, metrics and traces

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: exporter/elasticsearchexporter/model.go [hunk 1]
```diff
@@ -474,9 +482,21 @@ func (m *encodeModel) encodeAttributesOTelMode(document *objmodel.Document, attr
...
 	})
+	if stringifyArrayValues {
+		mapStringifyArrayValues(attributeMap)
+	}
 	document.AddAttributes("attributes", attributeMap)
 }
 
+func mapStringifyArrayValues(m pcommon.Map) {
+	m.Range(func(k string, v pcommon.Value) bool {
+		if v.Type() == pcommon.ValueTypeSlice {
```
Comment (felixbarny) [inline exporter/elasticsearchexporter/model.go:493]: This not only affects array values but also other complex attribute values, like `ValueTypeMap`. Not quite sure how to deal with `ValueTypeBytes` tbh. How do we serialize it normally? Is that compatible with a field type that [supports dimensions](https://www.elastic.co/guide/en/elasticsearch/reference/current/tsds.html#time-series-dimension)?

Comment (carsonip) [reply exporter/elasticsearchexporter/model.go:493]: according to https://opentelemetry.io/docs/specs/otel/common/ 
> The attribute value is either:
>
>    A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer.
>    An array of primitive type values. The array MUST be homogeneous, i.e., it MUST NOT contain values of different types.

I don't think ValueTypeMap and ValueTypeBytes fall into these valid value types.

Comment (felixbarny) [reply exporter/elasticsearchexporter/model.go:493]: How are we dealing with cases where we get invalid attribute values? Are we ignoring them?

Comment (carsonip) [reply exporter/elasticsearchexporter/model.go:493]: We had an issue in apm-server where otlp events with non-compliant attributes cause panics: https://github.com/elastic/apm-server/issues/13703

In es exporter, by looking at the code, ValueTypeMap will be successfully encoded as objects, ValueTypeBytes will be ignored. Ideally they should be dropped explicitly with a warning, but as these are non-compliant, user should expect undefined behavior, which includes successful or unsuccessful indexing.

Comment (carsonip) [reply exporter/elasticsearchexporter/model.go:493]: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35023

**Comment by carsonip** [comment]
Ideally https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35023 is addressed first, so that the impact of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35023 is not pronounced by this change.

File: exporter/elasticsearchexporter/model.go (hunk 1 above)
Comment (felixbarny) [reply exporter/elasticsearchexporter/model.go:516]: As long as we don't serialize non-compliant values and there's no panic, this sounds fine. Ideally, we would increment the dropped attributes count.

**Review by felixbarny — APPROVED** [review]

**Review by andrzej-stencel — APPROVED** [review]
:rocket:

**Pull request merged**